src/pip/_internal/pyproject.py

```diff
@@ -0,0 +1,144 @@
+from __future__ import absolute_import
+
+import io
+import os
+
+from pip._vendor import pytoml, six
+
+from pip._internal.exceptions import InstallationError
+
+
+def _is_list_of_str(obj):
+    return (
+        isinstance(obj, list) and
+        all(isinstance(item, six.string_types) for item in obj)
+    )
+
+
+def load_pyproject_toml(use_pep517, pyproject_toml, setup_py, req_name):
+    """Load the pyproject.toml file.
+
+    Parameters:
+        use_pep517 - Has the user requested PEP 517 processing? None
+                     means the user hasn't explicitly specified.
+        pyproject_toml - Location of the project's pyproject.toml file
+        setup_py - Location of the project's setup.py file
+        req_name - The name of the requirement we're processing (for
+                   error reporting)
+
+    Returns:
+        None if we should use the legacy code path, otherwise a tuple
+        (
+            requirements from pyproject.toml,
+            name of PEP 517 backend,
+            requirements we should check are installed after setting
+                up the build environment
+        )
+    """
+    has_pyproject = os.path.isfile(pyproject_toml)
+    has_setup = os.path.isfile(setup_py)
+
+    if has_pyproject:
+        with io.open(pyproject_toml, encoding="utf-8") as f:
+            pp_toml = pytoml.load(f)
+        build_system = pp_toml.get("build-system")
+    else:
+        build_system = None
+
+    # The following cases must use PEP 517
+    # We check for use_pep517 equalling False because that
+    # means the user explicitly requested --no-use-pep517
+    if has_pyproject and not has_setup:
+        if use_pep517 is False:
+            raise InstallationError(
+                "Disabling PEP 517 processing is invalid: "
+                "project does not have a setup.py"
+            )
+        use_pep517 = True
+    if build_system and "build-backend" in build_system:
+        if use_pep517 is False:
+            raise InstallationError(
+                "Disabling PEP 517 processing is invalid: "
+                "project specifies a build backend of {} "
+                "in pyproject.toml".format(
+                    build_system["build-backend"]
+                )
+            )
+        use_pep517 = True
+
+    # If we haven't worked out whether to use PEP 517 yet,
+    # and the user hasn't explicitly stated a preference,
+    # we do so if the project has a pyproject.toml file.
+    if use_pep517 is None:
+        use_pep517 = has_pyproject
+
+    if build_system is None and use_pep517:
```

Comment:
If you move this if-block to be after the:

```python
if not use_pep517:
    return None
```

then you can remove the "`and use_pep517`" from the conditional. I think it's also simpler / clearer this way because `build_system` isn't needed if you're returning with `None`.